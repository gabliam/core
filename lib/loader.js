"use strict";
const includeAll = require("include-all");
function loader(folder) {
    includeAll({
        dirname: folder,
        filter: /.*\.(js|ts)$/,
        excludeDirs: /^\.(git|svn|node_modules|dist|build)$/,
        identity: true
    });
}
exports.loader = loader;
