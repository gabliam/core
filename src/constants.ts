/** Type in registry */
export const TYPE = {
    Service: 'ServiceType',
    Config: 'ConfigType',
};

export const METADATA_KEY = {
    bean: '_bean',
    value: '_value',
    register: '_register',
    scan: '_scan',
};

export const ORDER_CONFIG = {
    Core: 50,
    Plugin: 100,
    Config: 150,
};

export const APP_CONFIG = Symbol('APP_CONFIG');

export const CORE_CONFIG = Symbol('CORE_CONFIG');
