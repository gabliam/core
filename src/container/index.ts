import { Container } from 'inversify';
import { makeLoggerMiddleware } from 'inversify-logger-middleware';
import { makeValueMiddleware } from './valueMiddleware';

/**
 * Create the inversify container
 */
export function createContainer(): Container {
    let container = new Container();

    let middlewares = [];
    if (process.env.NODE_ENV === 'development') {
        let logger = makeLoggerMiddleware();
        middlewares.push(logger);
    }

    middlewares.push(makeValueMiddleware(container));
    container.applyMiddleware(...middlewares);
    return container;
}

export { Container };
