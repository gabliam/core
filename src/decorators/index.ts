export * from './service';
export * from './bean';
export { Config, PluginConfig } from './config';
export * from './value';
export * from './scan';
export * from './register';
