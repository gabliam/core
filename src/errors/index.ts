export * from './value-validation-error';
