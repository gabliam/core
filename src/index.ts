import * as interfaces from './interfaces';
export * from './decorators';
export * from './gabliam';
export { APP_CONFIG, CORE_CONFIG } from './constants';

export * from './registry';
export * from './errors';

export {
  injectable, tagged, named, inject, optional, unmanaged, multiInject, targetName, decorate, interfaces as inversifyInterfaces
} from 'inversify';

export { interfaces };
