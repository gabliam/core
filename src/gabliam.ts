import * as inversify from 'inversify';
import * as interfaces from './interfaces';
import * as _ from 'lodash';
import {
  TYPE,
  METADATA_KEY,
  APP_CONFIG,
  CORE_CONFIG,
  VALUE_EXTRACTOR
} from './constants';
import { Loader } from './loader';
import { createContainer } from './container';
import { Registry } from './registry';
import * as d from 'debug';
import { configureValueExtractor } from './utils';

const debug = d('Gabliam:core');

const DEFAULT_CONFIG: interfaces.GabliamConfig = {
  scanPath: process.env.PWD!,
  configPath: process.env.GABLIAM_CONFIG_PATH || process.env.PWD!
};

/**
 * Gabliam
 */
export class Gabliam {
  public container: inversify.interfaces.Container = createContainer();

  public config: any;

  /**
   * Registry
   */
  public registry: Registry;

  private _loader: Loader = new Loader();

  private _plugins: interfaces.GabliamPlugin[] = [];
  private _options: interfaces.GabliamConfig;

  /**
   * Constructor
   * @param  {interfaces.GabliamConfig|string} options?
   */
  constructor(options?: Partial<interfaces.GabliamConfig> | string) {
    if (options === undefined) {
      this._options = DEFAULT_CONFIG;
    } else {
      if (_.isString(options)) {
        this._options = {
          scanPath: options,
          configPath: options
        };
      } else {
        this._options = {
          ...DEFAULT_CONFIG,
          ...options
        };
      }
    }
    /**
     * @TODO move in building phase
     */
    this.container.bind(CORE_CONFIG).toConstantValue(this._options);
  }
  /**
   * Add a plugin order
   * @param  {interfaces.GabliamPluginConstructor} ctor
   * @returns Gabliam
   */
  public addPlugin(ctor: interfaces.GabliamPluginConstructor): Gabliam {
    this._plugins.push(new ctor());
    return this;
  }

  /**
   * Build gabliam
   * @returns Promise
   */
  public async build(): Promise<Gabliam> {
    /**
     * Load config file
     */
    this._initializeConfig();

    /**
     * Loading phase
     */
    this.registry = this._loader.loadModules(
      this._options.scanPath,
      this._plugins
    );

    /**
     * Binding phase
     */
    this._bind();

    /**
     * Config phase
     */
    await this._loadConfig();

    /**
     * building phase
     */
    this._plugins
      .filter(plugin => _.isFunction(plugin.build))
      .forEach(plugin => plugin.build!(this.container, this.registry));

    return this;
  }

  /**
   * Build and start gabliam application
   * @returns Promise
   */
  async buildAndStart(): Promise<Gabliam> {
    await this.build();
    await this.start();
    return this;
  }

  /**
   * Start gabliam application
   *
   * call all plugin.start
   * @returns Promise
   */
  async start(): Promise<Gabliam> {
    const pluginsStart = this._plugins
      .filter(plugin => _.isFunction(plugin.start))
      .map(plugin => plugin.start!(this.container, this.registry));

    await Promise.all(pluginsStart);
    return this;
  }

  /**
   * Stop gabliam application
   * call all plugin.stop
   *
   * @returns Promise
   */
  async stop(): Promise<Gabliam> {
    const pluginsStop = this._plugins
      .filter(plugin => _.isFunction(plugin.stop))
      .map(plugin => plugin.stop!(this.container, this.registry));

    await Promise.all(pluginsStop);
    return this;
  }

  /**
   * Destroy gabliam application
   * call all plugin.destroy
   *
   * @returns Promise
   */
  async destroy(): Promise<Gabliam> {
    const pluginsDestroy = this._plugins
      .filter(plugin => _.isFunction(plugin.destroy))
      .map(plugin => plugin.destroy!(this.container, this.registry));

    await Promise.all(pluginsDestroy);
    return this;
  }

  /**
   * Load config file and bind result in APP_CONFIG
   */
  private _initializeConfig() {
    const config = (this.config = this._loader.loadConfig(
      this._options.configPath
    ));
    this.container.bind(APP_CONFIG).toConstantValue(config);
    this.container
      .bind(VALUE_EXTRACTOR)
      .toConstantValue(configureValueExtractor(this.container));
  }

  /**
   * Binding phase
   * Binding of config and service classes.
   * call all plugin.bind
   */
  private _bind() {
    debug('_bind');
    this.registry
      .get(TYPE.Config)
      .forEach(({ id, target }) =>
        this.container.bind(id).to(target).inSingletonScope()
      );

    this.registry
      .get(TYPE.Service)
      .forEach(({ id, target }) =>
        this.container.bind(id).to(target).inSingletonScope()
      );

    this._plugins
      .filter(plugin => _.isFunction(plugin.bind))
      .forEach(plugin => plugin.bind!(this.container, this.registry));
  }

  /**
   * Config phase
   */
  private async _loadConfig() {
    debug('_loadConfig');
    async function callInstance(instance: any, key: string) {
      return Promise.resolve(instance[key]());
    }

    // list of plugins with config phase
    const pluginConfig = this._plugins.filter(plugin =>
      _.isFunction(plugin.config)
    );

    let configsRegistry = this.registry.get<interfaces.ConfigRegistry>(
      TYPE.Config
    );

    debug('configsRegistry', configsRegistry);
    if (configsRegistry) {
      configsRegistry = _.sortBy(configsRegistry, 'order');

      for (const { id: configId } of configsRegistry) {
        const confInstance = this.container.get<object>(configId);

        const beanMetadata: interfaces.BeanMetadata[] = Reflect.getOwnMetadata(
          METADATA_KEY.bean,
          confInstance.constructor
        );

        if (beanMetadata) {
          // No promise.all and await because order of beans are important
          for (const metadata of beanMetadata) {
            this.container
              .bind(metadata.id)
              .toConstantValue(await callInstance(confInstance, metadata.key));
          }
        }

        pluginConfig.forEach(plugin =>
          plugin.config!(this.container, this.registry, confInstance)
        );
      }
    }
    debug('_loadConfig end');
  }
}
