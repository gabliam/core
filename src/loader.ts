import * as glob from 'glob';
import * as yaml from 'js-yaml';
import * as fs from 'fs';
import * as _ from 'lodash';
import * as d from 'debug';

import { Registry } from './registry';
import { hasMetadata } from './decorators';
import { METADATA_KEY } from './constants';
import { RegistryMetada, GabliamPlugin } from './interfaces';

const debug = d('Gabliam:loader');
const reg = /^.*(git|svn|node_modules|dist|build).*/;

/**
 * Loader
 */
export class Loader {
  /**
   * List of loaded folder
   */
  private loadedFolder: string[] = [];

  /**
   * Load all modules
   * @param  {string} scan
   * @param  {GabliamPlugin[]} plugins
   */
  loadModules(scan: string, plugins: GabliamPlugin[]) {
    const folders = plugins.reduce(
      (prev, current) => {
        if (hasMetadata(METADATA_KEY.scan, current.constructor)) {
          const paths = <string[]>Reflect.getOwnMetadata(
            METADATA_KEY.scan,
            current.constructor
          );
          prev.push(...paths);
        }
        return prev;
      },
      [scan]
    );
    debug('folders to load', folders);
    return this.loadFolders(...folders);
  }

  /**
   * Load on folder
   * @param  {string} folder
   */
  private loadFolder(folder: string) {
    debug(`load ${folder}`);

    const registry = new Registry();

    const files = glob
      .sync('**/*.@(js|ts)', { cwd: folder })
      .filter(file => !_.endsWith(file, '.d.ts') && !reg.test(file))
      .map(file => `${folder}/${file}`);

    for (const file of files) {
      const modules = require(file);
      for (const k of Object.keys(modules)) {
        const m = modules[k];
        if (hasMetadata(METADATA_KEY.register, m)) {
          const metadata = <RegistryMetada>Reflect.getOwnMetadata(
            METADATA_KEY.register,
            m
          );
          registry.add(metadata.type, metadata.value);
        }
        // if ()
        if (hasMetadata(METADATA_KEY.scan, m)) {
          const paths = <string[]>Reflect.getOwnMetadata(METADATA_KEY.scan, m);
          registry.addRegistry(this.loadFolders(...paths));
        }
      }
    }

    return registry;
  }

  /**
   * Load folders
   * @param  {string[]} ...folders
   */
  private loadFolders(...folders: string[]) {
    const registry = new Registry();

    if (!Array.isArray(folders) || folders.length === 0) {
      return registry;
    }

    for (const folder of folders) {
      if (this.loadedFolder.indexOf(folder) === -1) {
        this.loadedFolder.push(folder);
        registry.addRegistry(this.loadFolder(folder));
      }
    }
    return registry;
  }

  /**
   * Load configuration
   * @param  {string} folder the configuration folder
   * @returns any
   */
  loadConfig(folder: string): any {
    debug('loadConfig', folder);
    const files = glob.sync('**/application?(-+([a-zA-Z])).yml', {
      cwd: folder
    });
    let config = {};
    if (!files) {
      return config;
    }

    const profile = process.env.PROFILE || null;
    const defaultProfileFile = files.find(file => file === 'application.yml');

    if (defaultProfileFile) {
      config = this.loadYmlFile(`${folder}/${defaultProfileFile}`);
    }

    if (profile) {
      const profileFile = files.find(
        file => file === `application-${profile}.yml`
      );

      if (profileFile) {
        config = _.merge(
          {},
          config,
          this.loadYmlFile(`${folder}/${profileFile}`)
        );
      }
    }
    debug('loadConfig', config);
    return config;
  }

  private loadYmlFile(ymlPath: string) {
    const data = fs.readFileSync(ymlPath, 'utf8');
    try {
      return yaml.load(data) || {};
    } catch (e) {
      return {};
    }
  }
}
