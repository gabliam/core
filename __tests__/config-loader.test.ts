// tslint:disable:one-line
// tslint:disable:no-unused-expression
import { Loader } from '../src/loader';
import * as mock from 'mock-fs';

describe('Unit Test: config loader', () => {
  let loader: Loader;
  beforeEach(() => {
    loader = new Loader();
  });

  it(`with config folder doesn't exist`, () => {
    const config = loader.loadConfig('doesntexist/config');
    expect(config).toMatchSnapshot();
  });

  it(`with application.yml empty`, () => {
    mock({
      'test/config': {
        'application.yml': ``
      }
    });
    const config = loader.loadConfig('test/config');
    expect(config).toMatchSnapshot();
    mock.restore();
  });

  it(`with application.yml with nothing`, () => {
    mock({
      'test/config': {
        // tslint:disable-next-line:no-trailing-whitespace
        'application.yml': `

              `
      }
    });
    const config = loader.loadConfig('test/config');
    expect(config).toEqual({});
    mock.restore();
  });

  describe('with application.yml', () => {
    beforeAll(() => {
      mock({
        'test/config': {
          'application.yml': `
              application:
                host: 127.0.0.1
                db: test
            `,
          'application-prod.yml': `
              application:
                host: prod.app.com
            `
        }
      });
    });

    it('load application.yml', () => {
      const config = loader.loadConfig('test/config');
      expect(config).toMatchSnapshot();
    });

    it('load with bad profile', () => {
      const config = loader.loadConfig('test/config', 'int');
      expect(config).toMatchSnapshot();
    });

    it('load with prod profile', () => {
      const config = loader.loadConfig('test/config', 'prod');
      expect(config).toMatchSnapshot();
    });

    afterAll(() => mock.restore());
  });
});
