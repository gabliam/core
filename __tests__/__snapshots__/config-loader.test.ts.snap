// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with application.yml empty 1`] = `Object {}`;

exports[`with application.yml load application.yml 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`with application.yml load with bad profile 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "127.0.0.1",
  },
}
`;

exports[`with application.yml load with prod profile 1`] = `
Object {
  "application": Object {
    "db": "test",
    "host": "prod.app.com",
  },
}
`;

exports[`with config folder doesn't exist 1`] = `Object {}`;
